src/modules/statsd.c

```diff
@@ -492,3 +494,185 @@ noit_module_t statsd = {
   noit_statsd_initiate_check,
   NULL
 };
+
+static int statsd_tcp_handle_payload(noit_check_t *check, char *payload, size_t len) {
+  statsd_handle_payload(&check, 1, payload, len);
+  return 0;
+}
+static int statsd_tcp_handler(eventer_t e, int m, void *c, struct timeval *t) {
+  return listener_handler(e,m,c,t);
+}
+static int statsd_tcp_listener(eventer_t e, int m, void *c, struct timeval *t) {
+  return listener_listen_handler(e,m,c,t);
+}
+static int
+noit_statsd_tcp_initiate_check(noit_module_t *self,
+                               noit_check_t *check,
+                               int once, noit_check_t *cause) {
+  check->flags |= NP_PASSIVE_COLLECTION;
+  if (check->closure == NULL) {
+    listener_closure_t *ccl;
+    ccl =
+      listener_closure_alloc("statsd", self, check,
+                             nldeb, nlerr,
+                             statsd_tcp_handle_payload,
+                             NULL);
+    check->closure = ccl;
+    unsigned short port = 8126;
+    int rows_per_cycle = 100;
+    struct sockaddr_in skaddr;
+    int sockaddr_len;
+    const char *config_val;
+
+    if(mtev_hash_retr_str(check->config, "listen_port", strlen("listen_port"),
+                          (const char **)&config_val)) {
+      port = atoi(config_val);
+    }
+    ccl->port = port;
+
+    if(mtev_hash_retr_str(check->config, "rows_per_cycle",
+                          strlen("rows_per_cycle"),
+                          (const char **)&config_val)) {
+      rows_per_cycle = atoi(config_val);
+    }
+    ccl->rows_per_cycle = rows_per_cycle;
+
+    if(port > 0) ccl->ipv4_listen_fd = socket(AF_INET, NE_SOCK_CLOEXEC|SOCK_STREAM, IPPROTO_TCP);
+    if(ccl->ipv4_listen_fd < 0) {
+      if(port > 0) {
+        mtevL(noit_error, "statsd: socket failed: %s\n", strerror(errno));
+        return -1;
+      }
+    }
+    else {
+      if(eventer_set_fd_nonblocking(ccl->ipv4_listen_fd)) {
+        close(ccl->ipv4_listen_fd);
+        ccl->ipv4_listen_fd = -1;
+        mtevL(noit_error,
+              "statsd: could not set socket (IPv4) non-blocking: %s\n",
+              strerror(errno));
+        return -1;
+      }
+      socklen_t reuse = 1;
+      if (setsockopt(ccl->ipv4_listen_fd, SOL_SOCKET, SO_REUSEADDR, (void *)&reuse, sizeof(reuse)) != 0) {
+        mtevL(noit_error, "statsd listener(IPv4) failed(%s) to set REUSEADDR (doing our best)\n", strerror(errno));
+      }
+#ifdef SO_REUSEPORT
```

Comment:
If SO_REUSEPORT is defined, we should run the above code, AND the below code?  e.g. we call setsockops() twice?   This could be correct, but just looks weird to me.

Reply:
one is REUSEADDR (which is everywhere) the other is REUSEPORT (which is protected due to limited availability)